Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-                    word = ~word &+ 1
-                }
-
-                let digitsAdded = word.numericStringRepresentation(intoEndOfBuffer: &buffer[...wordInsertionPoint])
-
-                if .zero != quotient { // Not on the last word, so need to fill in leading zeroes etc.
-                    wordInsertionPoint -= decimalDigitsPerWord
-
-                    let leadingZeroes = decimalDigitsPerWord - digitsAdded
-                    assert(0 <= leadingZeroes, "Negative leading zeroes \(leadingZeroes)!  Expected \(decimalDigitsPerWord) digits per word and added \(digitsAdded).")
-
-                    if 0 < leadingZeroes {
-                        buffer[(wordInsertionPoint + 1)...(wordInsertionPoint + leadingZeroes)].initialize(repeating: UInt8(ascii: "0"))
-                    }
-                } else {
-                    wordInsertionPoint -= digitsAdded
-                }
-
-                tmp = quotient
-            }
-
-            if negative {
-                buffer[wordInsertionPoint] = UInt8(ascii: "-")
-                wordInsertionPoint -= 1
-            }
-
-            actualDigits = wordInsertionPoint.distance(to: buffer.endIndex) - 1
-
-            let unusedDigits = maximumDigits - actualDigits
-            assert(0 <= unusedDigits, "Negative unused digits \(unusedDigits)!  Expected at most \(maximumDigits) digit(s) but emitted \(actualDigits).")
-
-            if 0 < unusedDigits {
-                buffer[0..<unusedDigits].initialize(repeating: 0) // The buffer is permitted to be partially uninitialised, but only at the end.  So we have to initialise the unused portion at the start, in principle.  Technically this probably doesn't matter given we never subsequently read this part of the buffer, but there's no way to express that such that the compiler can ensure it for us.
-            }
+        numericStringRepresentationForBinaryInteger(words: self.words, isSigned: Self.isSigned)
+    }
+}
 
-            initialisedCount = maximumDigits
-        }[(maximumDigits - actualDigits)...]
+/// Formats `words` in "Numeric string" format (https://speleotrove.com/decimal/daconvs.html)
+/// which is the required input form for certain ICU functions (e.g. `unum_formatDecimal`).
+///
+/// - Parameters:
+///   - words: The binary integer's words.
+///   - isSigned: The binary integer's signedness.
+///
+internal // In order to test big integers without big integer models.
+func numericStringRepresentationForBinaryInteger(
+words: some Collection<UInt>, isSigned: Bool) -> ArraySlice<UInt8> {
+    // Copies the words and then passes them to a non-generic, mutating, word-based algorithm.
+    withUnsafeTemporaryAllocation(of: UInt.self, capacity: words.count) { copy in
```

Comment:
Hm. That is fair. I should measure some small values and perhaps retrofit the fast path you wrote. As a side note, it is possible to return a String instead of an ArraySlice through an additional temporary allocation. Given that String and withUnsafeTemporaryAllocation(\_:) can stack allocate, it might be beneficial. I haven't measured the difference, however.